f2_poly.c

```diff
@@ -5,112 +5,107 @@
 #include <fcntl.h>
 #include <string.h>
 #include <error.h>
+#include <math.h>
 #include "f2_poly.h"
 #include "arithm.h"
 
-
-size_t
-f2_poly_deg(f2_poly_t p){// La fonction qui calcule le degré d'un polynome.
-  size_t res = LEN-1 ; // Nous prenons un entier égal au degrès maximale que peut avoir un polynome
-  while(!((p>>res) & 1)){ // Une boucle qui permet de parcourir les bits
-    if(res<0)// jusqu'à arriver au premier bit non nul,
-      return -1 ;// Si le polynome est égale à 0, nous retournons -1
-    res-- ;
-  }
-  return res ;
+#ifdef __GNUC__
+/* clzll: GCC builtin function returning the number of leading-0 for unsigned long long */
+int8_t
+f2_poly_deg(f2_poly_t p){
+  return (p == ZERO) ? ZERO : __builtin_clzll(p)-1;
+}
+#else
+/* Computes the degree of a polynomial */
+/* by successively shifting the value representing the polynomial until the MSD is found */
+/* return -1 if the polynomial is null */
+int8_t
+f2_poly_deg(f2_poly_t p){
+  int8_t i = LEN-1 ;
+  if (p == 0x0) return -1;
+  while (NTH_BIT_IS_ZERO(p,i)) {i--;};
+  return i;
 }
 
-int
+void
 f2_poly_print(f2_poly_t p, const char c , FILE * fd){
-  int i ;
-  for(i = LEN -1; i>1 ; i-- ){ // Une boucle qui permet de parcourir les bits
-      if((p>>i) & 1){ // à chaque fois qu'un bit est égale à 1
-        fprintf(fd," + %c^%d ",c,i); // On l'affiche par le caractere donné argument
-      }
-  }
-  if(p&(1<<i)){ // le cas ou nous arrivons au monome X
-    fprintf(fd, " + %c ", c);
-  }
-  if(p&(1)){ // Le cas ou nous arrivons au monome constant
-    fprintf(fd," + 1 ");
-  }
-  fprintf(fd,"\n");// Pour plus de lisibilité, après un affichage, on revient a la ligne
-  return 0;
+  char fmt_xn[] = {" + %c^%d", ""};
```

Comment:
Reverse the array since we have an output if the bit is one
```suggestion
  char fmt_xn[] = {"", " + %c^%d"};
```